src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardInputActionObserver.cs

```diff
@@ -0,0 +1,406 @@
+            CreateKeyboardAction(inputActionMap, "minus");
+            CreateKeyboardAction(inputActionMap, "equals");
+            CreateKeyboardAction(inputActionMap, "backspace");
+            CreateKeyboardAction(inputActionMap, "insert");
+            CreateKeyboardAction(inputActionMap, "home");
+            CreateKeyboardAction(inputActionMap, "pageUp");
+            CreateKeyboardAction(inputActionMap, "numLock");
+            CreateKeyboardAction(inputActionMap, "numpadDivide");
+            CreateKeyboardAction(inputActionMap, "numpadMultiply");
+            CreateKeyboardAction(inputActionMap, "numpadMinus");
+
+            //row 3
+            CreateKeyboardAction(inputActionMap, "tab");
+            CreateKeyboardAction(inputActionMap, "q");
+            CreateKeyboardAction(inputActionMap, "w");
+            CreateKeyboardAction(inputActionMap, "e");
+            CreateKeyboardAction(inputActionMap, "r");
+            CreateKeyboardAction(inputActionMap, "t");
+            CreateKeyboardAction(inputActionMap, "y");
+            CreateKeyboardAction(inputActionMap, "u");
+            CreateKeyboardAction(inputActionMap, "i");
+            CreateKeyboardAction(inputActionMap, "o");
+            CreateKeyboardAction(inputActionMap, "p");
+            CreateKeyboardAction(inputActionMap, "leftBracket");
+            CreateKeyboardAction(inputActionMap, "rightBracket");
+            CreateKeyboardAction(inputActionMap, "backSlash");
+            CreateKeyboardAction(inputActionMap, "delete");
+            CreateKeyboardAction(inputActionMap, "end");
+            CreateKeyboardAction(inputActionMap, "pageDown");
+            CreateKeyboardAction(inputActionMap, "numpad7");
+            CreateKeyboardAction(inputActionMap, "numpad8");
+            CreateKeyboardAction(inputActionMap, "numpad9");
+            CreateKeyboardAction(inputActionMap, "numpadPlus");
+
+            //row 4
+            CreateKeyboardAction(inputActionMap, "capsLock");
+            CreateKeyboardAction(inputActionMap, "a");
+            CreateKeyboardAction(inputActionMap, "s");
+            CreateKeyboardAction(inputActionMap, "d");
+            CreateKeyboardAction(inputActionMap, "f");
+            CreateKeyboardAction(inputActionMap, "g");
+            CreateKeyboardAction(inputActionMap, "h");
+            CreateKeyboardAction(inputActionMap, "j");
+            CreateKeyboardAction(inputActionMap, "k");
+            CreateKeyboardAction(inputActionMap, "l");
+            CreateKeyboardAction(inputActionMap, "semicolon");
+            CreateKeyboardAction(inputActionMap, "quote");
+            CreateKeyboardAction(inputActionMap, "enter");
+            CreateKeyboardAction(inputActionMap, "numpad4");
+            CreateKeyboardAction(inputActionMap, "numpad5");
+            CreateKeyboardAction(inputActionMap, "numpad6");
+            // big plus button already in row 3 
+
+            //row 5
+            CreateKeyboardAction(inputActionMap, "leftShift");
```

Comment:
The beauty of this solution is that we don't care about the game's action bindings or input map when recording.

This records every key pressed, its hold duration, and its start time. so that if you pushed and held shift and then hit 1, we can tell that is different from holding 1 and pushing shift. The game may or may not treat those combinations differently, but we have captured them in their most pure state and will also replay them in that way.

I'm still working on the next PR after this one (replay) and I've been doing the mouse bits first... but for keyboard, I 'think' I have a way figured out to inject the raw keyboard events instead of having to bind them to actions on the inputactionassets.